proposals/000-cooperative-matrix-2.md

```diff
@@ -0,0 +1,524 @@
+
+An example of `decodeFunc` will be,
+```
+T decodeFunc(
+    const T* ptr,  // pointer to the raw data for the element
+    const uint32_t[Dim] blockCoord,   // block coordinates
+    const uint32_t[Dim] coordInBlock  // coordinates within the block
+);
+```
+
+**TODO**: It is unclear if we can use a raw memory pointer for the first parameter.
+
+```
+__generic<
+    let Dim : uint32_t,
+    let ClampMode : CoopMatClampMode>
+void store(
+    AnyBuffer buf,
+    uint elementOffset,
+    CoopMatTensorLayout<Dim, ClampMode> tensorLayout);
+
+__generic<
+    let Dim : uint32_t,
+    let ClampMode : CoopMatClampMode,
+    let DimView : uint32_t,
+    let HasDimensions : bool>
+void store(
+    AnyBuffer buf,
+    uint elementOffset,
+    CoopMatTensorLayout<Dim, ClampMode> tensorLayout,
+    CoopMatTensorView<DimView, HasDimensions> tensorView);
+```
+
+### Matrix Multiply-and-Accumulate operation
+
+`coopMatMulAdd()` performs a linear-algebraic matrix multiply of A by B and then component-wise add C.
+
+A, B, C and result matrices must use a same scope.
+
+The component types of A, B, C, result matrices don't need to be same.
+
+```
+__generic<
+    T : __BuiltinArithmeticType,
+    U : __BuiltinArithmeticType,
+    V : __BuiltinArithmeticType,
+    W : __BuiltinArithmeticType,
+    let S : CoopMatScope,
+    let M : int,
+    let K : int,
+    let N : int>
+CoopMat<T, S, M, N, CoopMatMatrixUse::MatrixAccumulator> coopMatMulAdd(
+    CoopMat<U, S, M, K, CoopMatMatrixUse::MatrixA> matA,
+    CoopMat<V, S, K, N, CoopMatMatrixUse::MatrixB> matB,
+    CoopMat<W, S, M, N, CoopMatMatrixUse::MatrixAccumulator> matC,
+    constexpr CoopMatMatrixOperands operands = CoopMatMatrixOperands::None);
+```
+
+### Reduce operation
+
+`coopMatReduce()` computes a matrix where each element of the result matrix is computed from a row, column, or neighborhood of the source matrix depending on `reduceMake`.
+
+```
+__generic<
+    T : __BuiltinFloatingPointType,
+    let S : CoopMatScope,
+    let M : int,
+    let N : int,
+    let RM : int,
+    let RN : int,
+    FuncType : IFunc<T, T, T>>
+void coopMatReduce(
+    out CoopMat<T, S, RM, RN, CoopMatMatrixUse::MatrixAccumulator> result,
+    CoopMat<T, S, M, N, CoopMatMatrixUse::MatrixAccumulator> m,
+    CoopMatReduceMask reduceMask,
+    DecodeFuncType combineOp);
+```
+
+`combineOp` is a user-defined function that implements `IFunc<T, T, T>`, which means the function returns a type `T` and it takes two arguments whose types are also `T`. In other words, the function takes two input values and reduces them to one value.
+
+### Per-Element operation
+
+Applies an operation to each element of a cooperative matrix.
+
+```
+__generic<
+    T : __BuiltinArithmeticType,
+    let S : CoopMatScope,
+    let M : int,
+    let N : int,
+    let R : CoopMatMatrixUse,
+    each UserDataTypes,
+    FuncType : IFunc<T, uint32_t, uint32_t, T, expand each UserDataTypes>>
+void coopMatPerElement(
```

Comment:
Let's not copy the GLSL design here. Instead of an `out` parameter, let's do
```
CoopMat<T,M,N> coopMatMapElement<T, int M, int N, each U>(
    expand CoopMat<each U, M, N> ms, IFunc<T, expand each U> f);
```